ls1intum/Artemis
# [Enhancement] Query improvements

### Checklist

### Motivation and Context


In this PR we try to replace some inefficient queries with more performant alternatives. Some queries are only re-formatted to be better readable.

### Description

The biggest part of this PR was to improve the performance of the assessment dashboard. There all the information about number of submissions, assessments and complaints was fetched individually from the database. 
This is partly refactored now. 
The number of submissions are fetched once, for all exercises.

Mostly I tried to improve the parformance of the following two REST calls, which I tested on develop with the PGDP course and with this branch.
on this branch:
![image](https://user-images.githubusercontent.com/33342534/112613581-8e27c200-8e20-11eb-8881-c3018bfdd603.png)

on develop:
![image](https://user-images.githubusercontent.com/33342534/112614677-d1366500-8e21-11eb-966d-7bb5ac70157c.png)

the `for-assessment-dashboard` call is now about twice as fast, and there is still the possibility of improvement. (I guess under 5s for PGdP could be possible)
the `stats-for-assessment-dashboard` call is still very slow, here is also still room for improvement, although it is more difficult.

### Steps for Testing

1. On the testServer you should check out different courses and see if the assessment dashboard works as before. Functionally there should be no difference.
2. If you have access to access to the production database test if the `for-assessment-dashboard` REST call is faster on this branch than on develop.

### Test coverage
stayed the same as only well tested classes are modified.
AssessmentDashboardService: 96%
ComplaintService:  87%
